ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaMain.cpp

```diff
@@ -21,37 +21,55 @@
 
 #include <unistd.h>
 #include <stdio.h>
+#include <mach-o/dyld.h>
+#include <limits.h>
 #include "CocoaMain.h"
 #include "ObjectiveC.h"
 #include <cstdlib>
+#include <string>
 
 #include "../General/PFwindow.h"
 #include "../General/PFfilemanip.h"
 
 #include "Platforms/General/PFmain.h"
 #include "Universal_System/roomsystem.h"
 
+using std::string;
+
 namespace enigma {
   void SetResizeFptr();
-}
+  
+  void initialize_directory_globals() {
+    // Set the working_directory
+    char buffer[PATH_MAX + 1];
+    if (getcwd(buffer, PATH_MAX + 1) != NULL)
+      enigma_user::working_directory = buffer;
 
-namespace enigma_user {
-  std::string working_directory = "";
+    // Set the program_directory
+    buffer[0] = 0;
+    char real_executable[PATH_MAX + 1];
+    char *bundle_id;
+
+    uint32_t bufsize = sizeof(buffer);
+    if (_NSGetExecutablePath(buffer, &bufsize) == 0) {
+      bundle_id = dirname(buffer);
+      strcpy(real_executable, bundle_id);
+      strcat(real_executable, "/");
+
+      enigma_user::program_directory = real_executable;
+    }
+
+    // Set the temp_directory
+    char const *env = getenv("TMPDIR");
+
+    if (env == 0)
+      env = "/tmp/";
+
+    enigma_user::temp_directory = env; 
+  }
 }
 
-extern "C" void copy_bundle_cwd(char* res);
-
-int main(int argc,char** argv)
-{
-  // Set the working_directory (from the bundle's location; using cwd won't work right on OS-X).
-  char buffer[1024] = {0};
-  copy_bundle_cwd(&buffer[0]);
-  if (buffer[0])
-    fprintf(stdout, "Current working dir: %s\n", buffer);
-  else
-    perror("copy_bundle_cwd() error");
-  enigma_user::working_directory = string( buffer );
-  
+int main(int argc,char** argv) {
```

Comment:
Nothing seems to call this in Cocoa. I don't see where `enigma_main` is reached, which is the method that invokes this logic, now. Are you sure this is being invoked? And if so, by what routine?

Reply:
Our Cocoa is totally screwed up now because of SDL and all of that stuff. It should be returning `enigma::enigma_main` from `main` but it isn't. Somebody needs to go in there and clean up that system. This is mostly unrelated to TKG here though, he's not responsible for it being so busted right now.